browser/brave_wallet/keyring_service_unittest.cc

```diff
@@ -3819,4 +3819,195 @@ TEST_F(KeyringServiceAccountDiscoveryUnitTest, RestoreWalletTwice) {
+        "YbQtaJQKLXET9jVjepWXe",
+        *GetPrivateKeyForKeyringAccount(
+            service, "C5ukMV73nk32h52MjxtnZXTrrr7rupD9CTDDRnYYDRYQ",
+            mojom::CoinType::SOL));
+
+    EXPECT_EQ(
+        "7b2254797065223a22736563703235366b31222c22507269766174"
+        "654b6579223a224169776f6a344469323155316844776835735348"
+        "434d7a37342b346c45303472376e5349454d706d6258493d227d",
+        *GetPrivateKeyForKeyringAccount(
+            service, "f1syhomjrwhjmavadwmrofjpiocb6r72h4qoy7ucq",
+            mojom::CoinType::FIL));
+
+    EXPECT_EQ(
+        "7b2254797065223a22736563703235366b31222c22507269766174"
+        "654b6579223a226376414367502f53344f3274796c4f42466a6348"
+        "33583154373677696661456c6646435057612b6a474a453d227d",
+        *GetPrivateKeyForKeyringAccount(
+            service, "t17puhwpgtnjr54kw7dwnjiphgn6kxlsyzbizwdhy",
+            mojom::CoinType::FIL));
+  }
+
+  const std::string& saved_mnemonic() const { return saved_mnemonic_; }
+
+ private:
+  std::string saved_mnemonic_;
+  base::test::ScopedFeatureList feature_list_;
+};
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest, NoMigration) {
+  EXPECT_TRUE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+  SetupKeyring();
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  KeyringService service(json_rpc_service(), GetPrefs());
+  EXPECT_TRUE(Unlock(&service, "brave"));
+  ValidateUnlockedKeyring(&service);
+}
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest, MigrateWithUnlock) {
+  // Setup prefs with legacy iterations count value.
+  KeyringService::GetPbkdf2IterationsForTesting() = 100000;
+  SetupKeyring();
+  KeyringService::GetPbkdf2IterationsForTesting() = absl::nullopt;
+
+  // Reset migration pref so migration runs on next unlock.
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+  GetPrefs()->ClearPref(kBraveWalletKeyringsNeedPasswordsMigration);
+  EXPECT_TRUE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  KeyringService service(json_rpc_service(), GetPrefs());
+  EXPECT_TRUE(Unlock(&service, "brave"));
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  service.Lock();
+  // Unlock with wrong password fails.
+  EXPECT_FALSE(Unlock(&service, "another password"));
+  KeyringService::GetPbkdf2IterationsForTesting() = 100000;
+  // Unlock with legacy iterations fails.
+  EXPECT_FALSE(Unlock(&service, "brave"));
+  KeyringService::GetPbkdf2IterationsForTesting() = absl::nullopt;
+
+  // Unlocking again works.
+  EXPECT_TRUE(Unlock(&service, "brave"));
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  ValidateUnlockedKeyring(&service);
+}
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest, MigrateWithRestore) {
```

Comment:
Normal restore doesn't need to run migration, it should be treated same as creating new wallet.
The code you are testing here is special path which is "Restore with current mnemonic" and what it does is not restore, it only does resume.

Reply:
Public interface for this is named as `RestoreWallet`, so I follow that name in tests.

Reply:
yep, I agree https://github.com/brave/brave-core/pull/14523/commits/af40a88c93cda80eb9bb36f283813c1858fdcc84